### PR TITLE
Em/feat/3Dcyl

### PR DESCRIPTION
**Pull request by lheagy**
This pr depends on simpeg/discretize#36 (3D cyl mesh and cell grad stencils being properties)

# Changes across EM
- electric fields / fluxes and magnetic fields fluxes live on the same grid location (eg. in problem_e, `e`, `j` are on edges, `b`, `h` are on faces

# TDEM improvements
- fill out the fields objects (addresses issue #633)
- added a raw vec grounded source for HJ forumlation

# DC
- a few typo fixes `Dirichlet` (some places it was  `Dirchlet`)
- a couple updates in the order of error checks to have the DC operate on the 3D cyl mesh

# General Cleanup
- import derivative tests from discretize

@sgkang, @rowanc1: Would you be willing to review?

### CODE REVIEW

**Comment by lheagy** [comment]
- [ ] add an example!

File: SimPEG/EM/Static/DC/FieldsDC.py [hunk 1]
```diff
@@ -76,17 +76,28 @@ class Fields_CC(FieldsDC):
...
-            if(self.prob.bc_type == 'Neumann'):
+
+        if getattr(self.survey.prob, 'bc_type', None) == 'Dirichlet':
+            self.cellGrad = -mesh.faceDiv.T
+        elif getattr(self.survey.prob, 'bc_type', None) == 'Neumann':
+            if self.mesh._meshType == "TREE":
                 raise NotImplementedError()
-            elif(self.prob.bc_type == 'Dirchlet'):
-                self.cellGrad = -mesh.faceDiv.T
+            mesh.setCellGradBC("neumann")
```
Comment (sgkang) [inline SimPEG/EM/Static/DC/FieldsDC.py]: Variable related cellGrad can be removed now. 
This was for computing electric field before.

File: SimPEG/EM/TDEM/ProblemTDEM.py [hunk 2]
```diff
@@ -1126,3 +1191,10 @@ def getRHS(self, tInd):
     def getRHSDeriv(self, tInd, src, v, adjoint=False):
         return Utils.Zero()  # assumes no derivs on sources
 
+    def getAdc(self):
```
Comment (sgkang) [inline SimPEG/EM/TDEM/ProblemTDEM.py]: Here we are assuming dirichlet b.c., so when considering air-earth interface we need to put air cells.
I guess that's fine for TDEM.

File: SimPEG/EM/TDEM/SrcTDEM.py [hunk 3]
```diff
@@ -577,3 +577,51 @@ def s_m(self, prob, time):
 
     def s_e(self, prob, time):
         return self.Mejs(prob) * self.waveform.eval(time)
+
+
+# TODO: this should be generalized and plugged into getting the Line current
```
Comment (sgkang) [inline SimPEG/EM/TDEM/SrcTDEM.py]: There is a line current implementation now!

File: SimPEG/Regularization.py [hunk 4]
```diff
@@ -411,7 +411,7 @@ def cellDiffxStencil(self):
         if getattr(self, '_cellDiffxStencil', None) is None:
 
             self._cellDiffxStencil = (
-                self.Pafx.T * self.mesh._cellGradxStencil() * self.Pac
+                self.Pafx.T * self.mesh._cellGradxStencil * self.Pac
```
Comment (sgkang) [inline SimPEG/Regularization.py]: so now `self.mesh._cellGradxStencil` is a property?

File: tests/em/static/test_DC_analytic.py [hunk 5]
```diff
@@ -144,10 +144,11 @@ def setUp(self):
         self.sigma = sigma
         self.data_anal = data_anal
 
-    def test_Problem3D_CC_Dirchlet(self, tolerance=0.2):
+    def test_Problem3D_CC_Dirichlet(self, tolerance=0.2):
```
Comment (sgkang) [inline tests/em/static/test_DC_analytic.py]: nice catch.

File: tests/em/tdem/test_TDEM_DerivAdjoint.py [hunk 6]
```diff
@@ -195,16 +195,34 @@ def test_Jvec_e_dbzdt(self):
...
+        def test_Jvec_e_jy(self):
+            self.JvecTest('jy')
+
     if testAdjoint:
-        def test_Jvec_adjoint_e_ey(self):
+        def test_Jvec_adjoint_e_dbdtx(self):
             self.JvecVsJtvecTest('dbdtx')
 
-        def test_Jvec_adjoint_e_ey(self):
+        def test_Jvec_adjoint_e_dbdtz(self):
```
Comment (sgkang) [inline tests/em/tdem/test_TDEM_DerivAdjoint.py]: Can we use dbdt for problem e now?

**Comment by lheagy** [comment]
This pr is superseded by #699

**Pull request closed without merging**